csharp_dotnetcore/3.Welcome-Users/WelcomeUserBot.cs

```diff
@@ -0,0 +1,146 @@
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = new CancellationToken())
+        {
+            // use state accessor to extract the didBotWelcomeUser flag
+            var didBotWelcomeUser = await _welcomeUserStateAccessors.DidBotWelcomedUser.GetAsync(turnContext, () => false);
+
+            if (turnContext.Activity.Type == ActivityTypes.Message && turnContext.Responded == false)
+            {
+                // Your bot should proactively send a welcome message to a personal chat the first time
+                // (and only the first time) a user initiates a personal chat with your bot.
+                if (didBotWelcomeUser == false)
+                {
+                    // update user state flag to reflect bot handled first user interaction.
+                    await _welcomeUserStateAccessors.DidBotWelcomedUser.SetAsync(turnContext, true);
+
+                    await turnContext.SendActivityAsync($"you first message ever to this bot ", cancellationToken: cancellationToken);
+                }
+                else
+                {
+                    var text = turnContext.Activity.Text.ToLowerInvariant();
+                    switch (text)
+                    {
+                        case "hello":
+                        case "hi":
+                            await turnContext.SendActivityAsync($"You said {text}.", cancellationToken: cancellationToken);
+                            await turnContext.SendActivityAsync(_genericMessage, cancellationToken: cancellationToken);
+                            break;
+                        case "intro":
+                        case "help":
+                            await SendIntroCardAsync(turnContext, cancellationToken);
+                            break;
+                        default:
+                            await turnContext.SendActivityAsync(_genericMessage, cancellationToken: cancellationToken);
+                            break;
+                    }
+                }
+            }
+
+            // greet when users are added to the conversation
+            else if (turnContext.Activity.Type == ActivityTypes.ConversationUpdate)
+            {
+                if (turnContext.Activity.MembersAdded.Any())
+                {
+                    // go over all new memebers, greet anyone that was not the target (recipient) of this message
+                    // the 'bot' is the recipient for events from the channel
+                    foreach (var member in turnContext.Activity.MembersAdded)
+                    {
+                        // turnContext.Activity.MembersAdded == turnContext.Activity.Recipient.Id indicates the
```

Comment:
Strongly consider a better comment. Remove the one on line 98, and put a comprehensive one here.